### PR TITLE
Add Keysystems protocol

### PR DESCRIPTION
**Pull request by teacupx**
The 'keysystems' protocol is used by some non-free dynamic DNS services such as www.domaindiscount24.com, www.moniker.com or www.rrpproxy.net.

Based on the work by Tom Bosmans (see https://www.gwbasics.be/blog.nsf/dx/custom-dynamic-dns-on-ubiquity-router-with-domaindiscount24.com.htm)

### CODE REVIEW

**Comment by SuperSandro2000** [comment]
I edited the post to remove the URL shorter link.

File: ddclient.in [hunk 1]
```diff
@@ -461,6 +461,10 @@ my %variables = (
         'static'              => setv(T_BOOL,  0, 1, 0,                    undef),
         'wildcard'            => setv(T_BOOL,  0, 1, 0,                    undef),
     },
+    'keysystems-common-defaults'       => {
+                       'server'              => setv(T_FQDNP,  1, 0, 1, 'dynamicdns.key-systems.net', undef),
+                       'login'               => setv(T_LOGIN,  0, 0, 0, 'unused',            undef),
```
Comment (SuperSandro2000) [inline ddclient.in:466]:
```suggestion
        'server'              => setv(T_FQDNP,  1, 0, 1, 'dynamicdns.key-systems.net', undef),
        'login'               => setv(T_LOGIN,  0, 0, 0, 'unused',            undef),
```

File: ddclient.in [hunk 2]
```diff
@@ -5787,6 +5800,83 @@ sub nic_gandi_update {
...
+o 'keysystems'
+
+The 'keysystems' protocol is used by some non-free dynamic DNS services
+such as www.domaindiscount24.com, www.moniker.com or www.rrpproxy.net.
+Check https://www.domaindiscount24.com/faq/en/dynamic-dns for API.
+
+Configuration variables applicable to the 'keysystems' protocol are:
+ protocol=keysystems               ##
+ server=www.fqdn.of.service   ## defaults to dynamicdns.key-systems.net
+ password=service-password    ## password (token) registered with the service
```
Comment (SuperSandro2000) [inline ddclient.in:5817]:
```suggestion
 server=www.example.com            ## defaults to dynamicdns.key-systems.net
 password=service-password         ## password (token) registered with the service
```

File: ddclient.in [hunk 3]
```diff
@@ -5787,6 +5800,83 @@ sub nic_gandi_update {
...
+ non-fully.qualified.host         ## the host registered with the service.
+
+Example ${program}.conf file entries:
+ ## single host update
+ protocol=keysystems,                                       \\
+ password=prettypassword                    \\
```
Comment (SuperSandro2000) [inline ddclient.in:5823]:
```suggestion
 protocol=keysystems,                       \\
 password=prettypassword                    \\
```

File: ddclient.in [hunk 4]
```diff
@@ -5787,6 +5800,83 @@ sub nic_gandi_update {
...
+
+       # Set the URL that we're going to to update
+       my $url;
+       $url  = "http://$config{$h}{'server'}/update.php";
+       $url .= "?hostname=";
+       $url .= $h;
+       $url .= "&password=";
+       $url .= $config{$h}{'password'};
+       $url .= "&ip=";
+       $url .= $ip;
```
Comment (SuperSandro2000) [inline ddclient.in:5853]:
```suggestion
       $url  = "http://$config{$h}{'server'}/update.php";
       $url .= "?hostname=$h";
       $url .= "&password=$config{$h}{'password'}";
       $url .= "&ip=$ip";
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: ddclient.in (hunk 4 above)
Comment (teacupx) [reply ddclient.in:5853]: Well, I tested it as it was, and it works. I cannot guarantee it will work with these changes. Are they really necessary?

Comment (rrthomas) [reply ddclient.in:5853]: FWIW, I have studied the above change, and it looks good to me; it simply reduces the number of lines of code by merging logically-related bits of string together.

File: ddclient.in [hunk 5]
```diff
@@ -5787,6 +5800,83 @@ sub nic_gandi_update {
...
+
+       if ($reply =~ /code = 200/)
+       {
+               $config{$h}{'ip'}     = $ip;
+               $config{$h}{'mtime'}  = $now;
+               $config{$h}{'status'} = 'good';
+               success("updating %s: good: IP address set to %s", $h, $ip);
+        }
+        else
+        {
```
Comment (SuperSandro2000) [inline ddclient.in:5873]:
```suggestion
        } else {
```
Please try to keep the additions short if possible. The file is already very big.

File: ddclient.in [hunk 6]
```diff
@@ -461,6 +461,10 @@ my %variables = (
         'static'              => setv(T_BOOL,  0, 1, 0,                    undef),
         'wildcard'            => setv(T_BOOL,  0, 1, 0,                    undef),
     },
+    'keysystems-common-defaults'       => {
+        'server'              => setv(T_FQDNP,  1, 0, 1, 'dynamicdns.key-systems.net', undef),
+        'login'               => setv(T_LOGIN,  0, 0, 0, 'unused',            undef),
+    },
```
[the remark this answers is not in the document]
Comment (teacupx) [reply ddclient.in:467]: It seems like a very reasonable and appropriate solution, for code uniformity and clarity. However, due to my limitations I am not sure how much will it take for me to be able to do it, I will try though.

**Comment by SuperSandro2000** [comment]
Closing due to inactivity. Feel free to comment and I can reopen the PR.

**Pull request closed without merging**